File: android/app/src/main/java/com/transcribecare/app/service/AudioPlayerService.kt
package com.transcribecare.app.service

import android.media.MediaPlayer
import android.media.PlaybackParams
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException

/**
 * Observable state representing the current audio player status.
 */
data class AudioPlayerState(
    val isPlaying: Boolean = false,
    val currentPosition: Int = 0,
    val totalDuration: Int = 0,
    val currentSpeed: Float = 1.0f,
    val error: String? = null
)

/**
 * Wrapper around Android's MediaPlayer that provides audio playback
 * with variable speed support and observable state via StateFlow.
 *
 * Supports playback speeds of 1x, 1.25x, 1.5x, and 2x.
 * Tracks progress position and total duration, and handles
 * end-of-playback by resetting to the beginning.
 *
 * @param scope CoroutineScope used for progress update polling.
 */
class AudioPlayerService(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main)
) {

    private var mediaPlayer: MediaPlayer? = null
    private var progressJob: Job? = null

    private val _state = MutableStateFlow(AudioPlayerState())
    val state: StateFlow<AudioPlayerState> = _state.asStateFlow()

    companion object {
        val SUPPORTED_SPEEDS = listOf(1.0f, 1.25f, 1.5f, 2.0f)
        private const val PROGRESS_UPDATE_INTERVAL_MS = 250L
    }

    /**
     * Starts playback of the audio file at the given path.
     *
     * If a file is already playing, it will be stopped and released before
     * starting the new file. Resets speed to 1x on new file load.
     *
     * @param filePath Absolute path to the audio file to play.
     */
    fun play(filePath: String) {
        // Validate file exists
        val file = File(filePath)
        if (!file.exists()) {
            _state.value = _state.value.copy(error = "Audio file not found.")
            return
        }

        // Release any existing player
        releasePlayer()

        try {
            val player = MediaPlayer().apply {
                setDataSource(filePath)
                prepare()
            }

            player.setOnCompletionListener {
                handlePlaybackComplete()
            }

            player.setOnErrorListener { _, what, extra ->
                val message = when (what) {
                    MediaPlayer.MEDIA_ERROR_IO -> "Error reading audio file."
                    MediaPlayer.MEDIA_ERROR_MALFORMED -> "Audio file format not supported."
                    MediaPlayer.MEDIA_ERROR_UNSUPPORTED -> "Audio format not supported."
                    else -> "An error occurred during playback."
                }
                _state.value = _state.value.copy(
                    isPlaying = false,
                    error = message
                )
                stopProgressUpdates()
                true
            }

            mediaPlayer = player

            val duration = player.duration
            _state.value = AudioPlayerState(
                isPlaying = true,
                currentPosition = 0,
                totalDuration = duration,
                currentSpeed = 1.0f,
                error = null
            )

            player.start()
            startProgressUpdates()
        } catch (e: IOException) {
            _state.value = _state.value.copy(error = "Unable to open audio file.")
        } catch (e: IllegalStateException) {
            _state.value = _state.value.copy(error = "Failed to initialize audio player.")
        } catch (e: IllegalArgumentException) {
            _state.value = _state.value.copy(error = "Invalid audio file path.")
        }
    }

    /**
     * Pauses the current playback. Has no effect if not currently playing.
     */
    fun pause() {
        val player = mediaPlayer ?: return
        if (!player.isPlaying) return

        try {
            player.pause()
            _state.value = _state.value.copy(
                isPlaying = false,
                currentPosition = player.currentPosition
            )
            stopProgressUpdates()
        } catch (e: IllegalStateException) {
            _state.value = _state.value.copy(error = "Failed to pause playback.")
        }
    }

    /**
     * Resumes playback from the current position. Has no effect if already playing.
     */
    fun resume() {
        val player = mediaPlayer ?: return
        if (player.isPlaying) return

        try {
            player.start()
            // Re-apply current speed after resuming
            applySpeed(_state.value.currentSpeed)
            _state.value = _state.value.copy(
                isPlaying = true,
                error = null
            )
            startProgressUpdates()
        } catch (e: IllegalStateException) {
            _state.value = _state.value.copy(error = "Failed to resume playback.")
        }
    }

    /**
     * Stops playback and resets position to the beginning.
     * The player remains prepared so play can be resumed via resume().
     */
    fun stop() {
        val player = mediaPlayer ?: return

        try {
            if (player.isPlaying) {
                player.pause()
            }
            player.seekTo(0)
            _state.value = _state.value.copy(
                isPlaying = false,
                currentPosition = 0
            )
            stopProgressUpdates()
        } catch (e: IllegalStateException) {
            _state.value = _state.value.copy(error = "Failed to stop playback.")
        }
    }

    /**
     * Sets the playback speed. Only values in [SUPPORTED_SPEEDS] are accepted.
     *
     * @param speed The desired playback speed (1.0f, 1.25f, 1.5f, or 2.0f).
     */
    fun setSpeed(speed: Float) {
        if (speed !in SUPPORTED_SPEEDS) return

        _state.value = _state.value.copy(currentSpeed = speed)
        applySpeed(speed)
    }

    /**
     * Seeks to a specific position in the audio.
     *
     * @param positionMs The target position in milliseconds.
     */
    fun seekTo(positionMs: Int) {
        val player = mediaPlayer ?: return
        val clampedPosition = positionMs.coerceIn(0, _state.value.totalDuration)

        try {
            player.seekTo(clampedPosition)
            _state.value = _state.value.copy(currentPosition = clampedPosition)
        } catch (e: IllegalStateException) {
            _state.value = _state.value.copy(error = "Failed to seek.")
        }
    }

    /**
     * Releases all resources held by the MediaPlayer.
     * Should be called when the service is no longer needed.
     * After calling release(), a new play() call is required to use the service again.
     */
    fun release() {
        stopProgressUpdates()
        releasePlayer()
        _state.value = AudioPlayerState()
    }

    /**
     * Clears the current error state.
     */
    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }

    private fun handlePlaybackComplete() {
        val player = mediaPlayer ?: return
        try {
            player.seekTo(0)
        } catch (_: IllegalStateException) {
            // Ignore seek errors on completion
        }
        _state.value = _state.value.copy(
            isPlaying = false,
            currentPosition = 0
        )
        stopProgressUpdates()
    }

    private fun applySpeed(speed: Float) {
        val player = mediaPlayer ?: return
        try {
            val params = PlaybackParams().setSpeed(speed)
            player.playbackParams = params
        } catch (_: IllegalStateException) {
            // Speed change not possible in current state — ignore silently
        } catch (_: IllegalArgumentException) {
            // Invalid speed value — ignore silently
        }
    }

    private fun startProgressUpdates() {
        stopProgressUpdates()
        progressJob = scope.launch {
            while (isActive) {
                val player = mediaPlayer
                if (player != null && player.isPlaying) {
                    _state.value = _state.value.copy(
                        currentPosition = player.currentPosition
                    )
                }
                delay(PROGRESS_UPDATE_INTERVAL_MS)
            }
        }
    }

    private fun stopProgressUpdates() {
        progressJob?.cancel()
        progressJob = null
    }

    private fun releasePlayer() {
        try {
            mediaPlayer?.release()
        } catch (_: Exception) {
            // Ignore release errors
        }
        mediaPlayer = null
    }
}

File: ios/TranscribeCare/Service/AudioPlayerService.swift
import Foundation
import AVFoundation

/// Wrapper around AVAudioPlayer that provides audio playback with variable
/// speed support, progress tracking, and end-of-playback handling.
///
/// Supports playback speeds of 1x, 1.25x, 1.5x, and 2x via the AVAudioPlayer
/// `rate` property. Exposes observable state for current position, total duration,
/// playing status, and current speed.
///
/// - Requirements: 5.3, 5.4, 5.5
@Observable
class AudioPlayerService: NSObject, AVAudioPlayerDelegate {

    // MARK: - Public State

    /// Current playback position in seconds.
    private(set) var currentPosition: TimeInterval = 0

    /// Total duration of the loaded audio file in seconds.
    private(set) var totalDuration: TimeInterval = 0

    /// Whether audio is currently playing.
    private(set) var isPlaying: Bool = false

    /// The current playback speed (1.0, 1.25, 1.5, or 2.0).
    private(set) var currentSpeed: Float = 1.0

    /// User-facing error message, or nil if no error.
    private(set) var error: String?

    // MARK: - Callbacks

    /// Called with a user-facing error message when playback encounters a problem.
    var onError: ((String) -> Void)?

    // MARK: - Constants

    /// Supported playback speed values.
    static let supportedSpeeds: [Float] = [1.0, 1.25, 1.5, 2.0]

    /// Interval between progress updates in seconds.
    private static let progressUpdateInterval: TimeInterval = 0.25

    // MARK: - Private Properties

    private var audioPlayer: AVAudioPlayer?
    private var progressTimer: Timer?

    // MARK: - Public Methods

    /// Starts playback of the audio file at the given URL.
    ///
    /// If audio is already playing, it will be stopped before starting the new file.
    /// Resets speed to 1x on new file load.
    ///
    /// - Parameter url: The file URL of the audio to play.
    func play(url: URL) {
        // Validate file exists
        guard FileManager.default.fileExists(atPath: url.path) else {
            let message = "Audio file not found."
            error = message
            onError?(message)
            return
        }

        // Release any existing player
        stopProgressUpdates()
        audioPlayer?.stop()
        audioPlayer = nil

        do {
            let player = try AVAudioPlayer(contentsOf: url)
            player.delegate = self
            player.enableRate = true
            player.rate = 1.0
            player.prepareToPlay()

            audioPlayer = player
            totalDuration = player.duration
            currentPosition = 0
            currentSpeed = 1.0
            error = nil

            guard player.play() else {
                let message = "Failed to start audio playback."
                error = message
                onError?(message)
                return
            }

            isPlaying = true
            startProgressUpdates()
        } catch let nsError as NSError {
            handlePlayerError(nsError)
        }
    }

    /// Pauses the current playback. Has no effect if not currently playing.
    func pause() {
        guard let player = audioPlayer, player.isPlaying else { return }

        player.pause()
        isPlaying = false
        currentPosition = player.currentTime
        stopProgressUpdates()
    }

    /// Resumes playback from the current position. Has no effect if already playing.
    func resume() {
        guard let player = audioPlayer, !player.isPlaying else { return }

        player.rate = currentSpeed
        guard player.play() else {
            let message = "Failed to resume playback."
            error = message
            onError?(message)
            return
        }

        isPlaying = true
        error = nil
        startProgressUpdates()
    }

    /// Stops playback and resets position to the beginning.
    func stop() {
        guard let player = audioPlayer else { return }

        player.stop()
        player.currentTime = 0
        isPlaying = false
        currentPosition = 0
        stopProgressUpdates()
    }

    /// Sets the playback speed. Only values in `supportedSpeeds` are accepted.
    ///
    /// The speed change takes effect immediately if audio is currently playing.
    ///
    /// - Parameter speed: The desired playback speed (1.0, 1.25, 1.5, or 2.0).
    func setSpeed(speed: Float) {
        guard Self.supportedSpeeds.contains(speed) else { return }

        currentSpeed = speed

        if let player = audioPlayer, player.isPlaying {
            player.rate = speed
        }
    }

    /// Releases all resources held by the audio player.
    /// Should be called when the service is no longer needed.
    func release() {
        stopProgressUpdates()
        audioPlayer?.stop()
        audioPlayer?.delegate = nil
        audioPlayer = nil
        isPlaying = false
        currentPosition = 0
        totalDuration = 0
        currentSpeed = 1.0
        error = nil
    }

    /// Clears the current error state.
    func clearError() {
        error = nil
    }

    // MARK: - AVAudioPlayerDelegate

    /// Called when audio playback finishes. Resets to beginning and stops.
    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        player.currentTime = 0
        isPlaying = false
        currentPosition = 0
        stopProgressUpdates()
    }

    /// Called when a decode error occurs during playback.
    func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error decodeError: (any Error)?) {
        let message = "Audio format not supported."
        self.error = message
        onError?(message)
        isPlaying = false
        currentPosition = 0
        stopProgressUpdates()
    }

    // MARK: - Private Methods

    /// Starts a repeating timer that updates the current playback position.
    private func startProgressUpdates() {
        stopProgressUpdates()
        progressTimer = Timer.scheduledTimer(
            withTimeInterval: Self.progressUpdateInterval,
            repeats: true
        ) { [weak self] _ in
            self?.updateProgress()
        }
    }

    /// Stops the progress update timer.
    private func stopProgressUpdates() {
        progressTimer?.invalidate()
        progressTimer = nil
    }

    /// Updates the current position from the audio player.
    private func updateProgress() {
        guard let player = audioPlayer, player.isPlaying else { return }
        currentPosition = player.currentTime
    }

    /// Handles errors from AVAudioPlayer initialization.
    private func handlePlayerError(_ nsError: NSError) {
        let message: String

        switch nsError.domain {
        case NSOSStatusErrorDomain:
            message = "Audio format not supported."
        case NSCocoaErrorDomain where nsError.code == NSFileReadNoSuchFileError:
            message = "Audio file not found."
        case NSCocoaErrorDomain where nsError.code == NSFileReadCorruptFileError:
            message = "Audio file is corrupted or in an unsupported format."
        default:
            message = "Unable to open audio file: \(nsError.localizedDescription)"
        }

        error = message
        onError?(message)
        isPlaying = false
        currentPosition = 0
        totalDuration = 0
    }
}
